File: packages/attrition-sdk/attrition/advisor.py
"""Advisor Mode tracking for attrition.

Instruments the Sonnet executor + Opus advisor pattern with real cost tracking.
Detects escalation events, measures advisor effectiveness, and pushes structured
packets to the attrition backend.

Usage:
    from attrition.advisor import AdvisorTracker

    tracker = AdvisorTracker(
        executor_model="claude-sonnet-4-6",
        advisor_model="claude-opus-4-6",
        endpoint="https://attrition-7xtb75zi5q-uc.a.run.app",
    )

    # Track an executor call
    tracker.log_executor_call(input_tokens=1200, output_tokens=400, tool="Edit", success=True)

    # Track an escalation to advisor
    tracker.log_advisor_call(
        trigger="executor_failure",
        input_tokens=3200,
        output_tokens=1800,
        advice_type="diagnosis",
        advice_summary="Race condition in WebSocket reconnect",
    )

    # End session and push summary
    tracker.end_session(task_completed=True, user_corrections=0)
"""

import json
import time
import urllib.request
from dataclasses import dataclass, field
from typing import Optional
from datetime import datetime, timezone

from attrition.storage import append_event

# ── Model pricing (USD per million tokens) ──────────────────────────────────

MODEL_PRICING: dict[str, tuple[float, float]] = {
    "claude-opus-4-6":   (15.0,  75.0),
    "claude-sonnet-4-6": (3.0,   15.0),
    "claude-haiku-4-5":  (0.8,   4.0),
    "gpt-4o":            (2.5,   10.0),
    "gpt-4o-mini":       (0.15,  0.6),
    "gemini-3.1-flash-lite-preview": (0.075, 0.30),
    "gemini-2.5-flash":  (0.15,  0.60),
}


def compute_cost(model: str, input_tokens: int, output_tokens: int) -> float:
    """Compute real cost from token counts using published pricing."""
    inp_rate, out_rate = MODEL_PRICING.get(model, (3.0, 15.0))
    return (input_tokens / 1_000_000) * inp_rate + (output_tokens / 1_000_000) * out_rate


# ── Data classes ────────────────────────────────────────────────────────────

@dataclass
class ExecutorCall:
    input_tokens: int
    output_tokens: int
    cost_usd: float
    tool: str
    success: bool
    timestamp: str


@dataclass
class AdvisorCall:
    trigger: str
    input_tokens: int
    output_tokens: int
    cost_usd: float
    advice_type: str
    advice_summary: str
    latency_ms: int
    was_applied: bool
    timestamp: str


@dataclass
class SessionStats:
    executor_tokens: int = 0
    executor_cost: float = 0.0
    executor_calls: int = 0
    advisor_tokens: int = 0
    advisor_cost: float = 0.0
    advisor_calls: int = 0
    escalation_triggers: list = field(default_factory=list)
    advice_types: list = field(default_factory=list)


# ── Main tracker ────────────────────────────────────────────────────────────

class AdvisorTracker:
    """Track Sonnet executor + Opus advisor pattern with real cost measurement."""

    def __init__(
        self,
        executor_model: str = "claude-sonnet-4-6",
        advisor_model: str = "claude-opus-4-6",
        endpoint: Optional[str] = None,
        session_id: Optional[str] = None,
        auto_push: bool = True,
    ):
        self.executor_model = executor_model
        self.advisor_model = advisor_model
        self.endpoint = endpoint or "https://attrition-7xtb75zi5q-uc.a.run.app"
        self.session_id = session_id or f"sess_{int(time.time() * 1000) % 1_000_000_000:x}"
        self.auto_push = auto_push
        self.started_at = datetime.now(timezone.utc).isoformat()

        self._executor_calls: list[ExecutorCall] = []
        self._advisor_calls: list[AdvisorCall] = []
        self._stats = SessionStats()

    def log_executor_call(
        self,
        input_tokens: int,
        output_tokens: int,
        tool: str = "",
        success: bool = True,
    ) -> ExecutorCall:
        """Record an executor (Sonnet) LLM call."""
        cost = compute_cost(self.executor_model, input_tokens, output_tokens)
        call = ExecutorCall(
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cost_usd=cost,
            tool=tool,
            success=success,
            timestamp=datetime.now(timezone.utc).isoformat(),
        )
        self._executor_calls.append(call)
        self._stats.executor_tokens += input_tokens + output_tokens
        self._stats.executor_cost += cost
        self._stats.executor_calls += 1

        # Log to local JSONL
        append_event({
            "ts": call.timestamp,
            "type": "advisor.executor_call",
            "session_id": self.session_id,
            "model": self.executor_model,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "cost_usd": round(cost, 8),
            "tool": tool,
            "success": success,
        })
        return call

    def log_advisor_call(
        self,
        trigger: str,
        input_tokens: int,
        output_tokens: int,
        advice_type: str = "general",
        advice_summary: str = "",
        latency_ms: int = 0,
        was_applied: bool = True,
    ) -> AdvisorCall:
        """Record an advisor (Opus) escalation call."""
        cost = compute_cost(self.advisor_model, input_tokens, output_tokens)
        call = AdvisorCall(
            trigger=trigger,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cost_usd=cost,
            advice_type=advice_type,
            advice_summary=advice_summary,
            latency_ms=latency_ms,
            was_applied=was_applied,
            timestamp=datetime.now(timezone.utc).isoformat(),
        )
        self._advisor_calls.append(call)
        self._stats.advisor_tokens += input_tokens + output_tokens
        self._stats.advisor_cost += cost
        self._stats.advisor_calls += 1
        self._stats.escalation_triggers.append(trigger)
        self._stats.advice_types.append(advice_type)

        # Log to local JSONL
        append_event({
            "ts": call.timestamp,
            "type": "advisor.advisor_call",
            "session_id": self.session_id,
            "model": self.advisor_model,
            "trigger": trigger,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "cost_usd": round(cost, 8),
            "advice_type": advice_type,
            "advice_summary": advice_summary[:200],
            "was_applied": was_applied,
        })

        # Push decision packet to backend
        if self.auto_push:
            self._push_decision(call)

        return call

    def end_session(
        self,
        task_completed: bool = True,
        user_corrections: int = 0,
        subject: str = "",
    ) -> dict:
        """Finalize the session and push summary packet."""
        ended_at = datetime.now(timezone.utc).isoformat()
        s = self._stats
        combined_cost = s.executor_cost + s.advisor_cost
        total_calls = s.executor_calls + s.advisor_calls

        # Estimate Opus-only cost (all tokens at Opus pricing)
        total_tokens = s.executor_tokens + s.advisor_tokens
        opus_only = compute_cost("claude-opus-4-6", total_tokens // 2, total_tokens // 2)

        session_data = {
            "session_id": self.session_id,
            "started_at": self.started_at,
            "ended_at": ended_at,
            "duration_ms": int(
                (datetime.fromisoformat(ended_at) - datetime.fromisoformat(self.started_at)).total_seconds() * 1000
            ),
            "executor_model": self.executor_model,
            "advisor_model": self.advisor_model,
            "executor_stats": {
                "total_tokens": s.executor_tokens,
                "total_cost_usd": round(s.executor_cost, 8),
                "calls": s.executor_calls,
                "self_sufficient_pct": round(
                    (s.executor_calls / total_calls * 100) if total_calls > 0 else 100, 1
                ),
            },
            "advisor_stats": {
                "total_tokens": s.advisor_tokens,
                "total_cost_usd": round(s.advisor_cost, 8),
                "calls": s.advisor_calls,
                "escalation_triggers": list(set(s.escalation_triggers)),
                "advice_types": list(set(s.advice_types)),
            },
            "combined": {
                "total_cost_usd": round(combined_cost, 8),
                "advisor_cost_share_pct": round(
                    (s.advisor_cost / combined_cost * 100) if combined_cost > 0 else 0, 1
                ),
                "escalation_rate_pct": round(
                    (s.advisor_calls / total_calls * 100) if total_calls > 0 else 0, 1
                ),
                "user_corrections": user_corrections,
                "task_completed": task_completed,
                "first_pass_success": task_completed and user_corrections == 0,
            },
            "comparison": {
                "opus_only_estimated_tokens": total_tokens,
                "opus_only_estimated_cost_usd": round(opus_only, 4),
                "savings_vs_opus_only_pct": round(
                    ((opus_only - combined_cost) / opus_only * 100) if opus_only > 0 else 0, 1
                ),
            },
        }

        # Log to local JSONL
        append_event({
            "ts": ended_at,
            "type": "advisor.session_end",
            "session_id": self.session_id,
            **{k: v for k, v in session_data.items() if k != "session_id"},
        })

        # Push session packet to backend
        if self.auto_push:
            self._push_session(session_data, subject)

        return session_data

    # ── Push helpers ─────────────────────────────────────────────────────

    def _push_decision(self, call: AdvisorCall):
        """Push advisor.decision packet to attrition backend (fire-and-forget)."""
        try:
            packet = {
                "type": "advisor.decision",
                "subject": f"Escalation: {call.advice_type} ({call.trigger})",
                "summary": call.advice_summary[:200] or f"{call.trigger} -> {call.advice_type}",
                "data": {
                    "session_id": self.session_id,
                    "decision_id": f"dec_{int(time.time() * 1000) % 1_000_000:x}",
                    "timestamp": call.timestamp,
                    "trigger": call.trigger,
                    "advisor": {
                        "model": self.advisor_model,
                        "input_tokens": call.input_tokens,
                        "output_tokens": call.output_tokens,
                        "total_tokens": call.input_tokens + call.output_tokens,
                        "cost_usd": round(call.cost_usd, 8),
                        "latency_ms": call.latency_ms,
                        "advice_type": call.advice_type,
                        "advice_summary": call.advice_summary[:500],
                    },
                    "outcome": {
                        "executor_applied": call.was_applied,
                        "task_completed": False,  # Unknown at decision time
                    },
                },
            }
            self._push_packet(packet)
        except Exception:
            pass  # Never crash the host

    def _push_session(self, data: dict, subject: str):
        """Push advisor.session packet to attrition backend (fire-and-forget)."""
        try:
            s = self._stats
            cost_str = f"${data['combined']['total_cost_usd']:.4f}"
            esc = s.advisor_calls
            packet = {
                "type": "advisor.session",
                "subject": subject or f"Session: {self.session_id}",
                "summary": f"{esc} escalations, {data['combined']['advisor_cost_share_pct']}% advisor cost, "
                           f"total {cost_str}",
                "data": data,
            }
            self._push_packet(packet)
        except Exception:
            pass

    def _push_packet(self, packet: dict):
        """POST to /api/retention/push-packet."""
        try:
            body = json.dumps(packet).encode("utf-8")
            req = urllib.request.Request(
                f"{self.endpoint}/api/retention/push-packet",
                data=body,
                headers={"Content-Type": "application/json"},
                method="POST",
            )
            urllib.request.urlopen(req, timeout=5)
        except Exception:
            pass  # Best-effort, never block the agent

File: rust/crates/api/src/routes/advisor.rs
//! Advisor Mode analytics endpoints.
//!
//! Computes aggregate stats from advisor.* retention packets.
//! No new storage — reads from existing retention_packets in AppState.
//!
//! GET /api/advisor/stats     — Aggregate advisor mode statistics
//! GET /api/advisor/sessions  — Per-session advisor summaries
//! GET /api/advisor/decisions — Individual escalation decisions

use axum::{
    extract::State,
    routing::get,
    Json, Router,
};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use crate::state::AppState;

// ── Model pricing (USD per million tokens) ─────────────────────────────────

const PRICING: &[(&str, f64, f64)] = &[
    ("claude-opus-4-6",   15.0,  75.0),
    ("claude-sonnet-4-6",  3.0,  15.0),
    ("claude-haiku-4-5",   0.8,   4.0),
    ("gpt-4o",             2.5,  10.0),
    ("gpt-4o-mini",        0.15,  0.6),
    ("gemini-3.1-flash-lite-preview", 0.075, 0.30),
    ("gemini-2.5-flash",   0.15,  0.60),
];

fn model_cost(model: &str, input_tokens: u64, output_tokens: u64) -> f64 {
    let (inp_rate, out_rate) = PRICING
        .iter()
        .find(|(name, _, _)| model.contains(name))
        .map(|(_, i, o)| (*i, *o))
        .unwrap_or((3.0, 15.0)); // default to Sonnet pricing
    (input_tokens as f64 / 1_000_000.0) * inp_rate
        + (output_tokens as f64 / 1_000_000.0) * out_rate
}

// ── Response types ──────────────────────────────────────────────────────────

#[derive(Serialize)]
struct AdvisorStats {
    total_sessions: usize,
    total_decisions: usize,
    total_pipeline_runs: usize,

    // Cost breakdown
    executor_total_cost_usd: f64,
    advisor_total_cost_usd: f64,
    combined_total_cost_usd: f64,
    advisor_cost_share_pct: f64,

    // Token breakdown
    executor_total_tokens: u64,
    advisor_total_tokens: u64,

    // Effectiveness
    escalation_rate_pct: f64,
    advisor_resolved_pct: f64,
    avg_user_corrections: f64,

    // Comparison
    estimated_opus_only_cost_usd: f64,
    savings_vs_opus_only_pct: f64,

    // Model usage
    models_seen: Vec<String>,

    // Per-model breakdown
    model_breakdown: Vec<ModelBreakdown>,
}

#[derive(Serialize)]
struct ModelBreakdown {
    model: String,
    calls: u64,
    total_tokens: u64,
    total_cost_usd: f64,
    avg_tokens_per_call: u64,
    role: String, // "executor" | "advisor" | "pipeline"
}

#[derive(Serialize)]
struct AdvisorSession {
    session_id: String,
    subject: String,
    executor_model: String,
    advisor_model: String,
    executor_tokens: u64,
    executor_cost_usd: f64,
    advisor_tokens: u64,
    advisor_cost_usd: f64,
    escalation_count: u64,
    task_completed: bool,
    user_corrections: u64,
    timestamp: String,
}

#[derive(Serialize)]
struct AdvisorDecision {
    decision_id: String,
    session_id: String,
    trigger: String,
    advisor_model: String,
    advisor_tokens: u64,
    advisor_cost_usd: f64,
    advice_type: String,
    advice_summary: String,
    was_applied: bool,
    timestamp: String,
}

#[derive(Serialize)]
struct StatsResponse {
    stats: AdvisorStats,
    message: String,
}

#[derive(Serialize)]
struct SessionsResponse {
    sessions: Vec<AdvisorSession>,
}

#[derive(Serialize)]
struct DecisionsResponse {
    decisions: Vec<AdvisorDecision>,
}

// ── Helpers ─────────────────────────────────────────────────────────────────

fn extract_str(data: &serde_json::Value, key: &str) -> String {
    data.get(key)
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string()
}

fn extract_u64(data: &serde_json::Value, key: &str) -> u64 {
    data.get(key)
        .and_then(|v| v.as_u64())
        .unwrap_or(0)
}

fn extract_f64(data: &serde_json::Value, key: &str) -> f64 {
    data.get(key)
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0)
}

fn extract_bool(data: &serde_json::Value, key: &str) -> bool {
    data.get(key)
        .and_then(|v| v.as_bool())
        .unwrap_or(false)
}

fn nested_str(data: &serde_json::Value, outer: &str, inner: &str) -> String {
    data.get(outer)
        .and_then(|v| v.get(inner))
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string()
}

fn nested_u64(data: &serde_json::Value, outer: &str, inner: &str) -> u64 {
    data.get(outer)
        .and_then(|v| v.get(inner))
        .and_then(|v| v.as_u64())
        .unwrap_or(0)
}

fn nested_f64(data: &serde_json::Value, outer: &str, inner: &str) -> f64 {
    data.get(outer)
        .and_then(|v| v.get(inner))
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0)
}

// ── Handlers ────────────────────────────────────────────────────────────────

async fn stats(
    State(state): State<Arc<AppState>>,
) -> Json<StatsResponse> {
    state.increment_requests();

    let packets = state.retention_packets.lock().await;

    // Separate packet types
    let decisions: Vec<_> = packets
        .iter()
        .filter(|p| p.packet_type == "advisor.decision")
        .collect();
    let sessions: Vec<_> = packets
        .iter()
        .filter(|p| p.packet_type == "advisor.session")
        .collect();
    let pipeline_runs: Vec<_> = packets
        .iter()
        .filter(|p| p.packet_type == "delta.pipeline_run")
        .collect();

    // Aggregate from advisor.session packets
    let mut executor_cost = 0.0_f64;
    let mut advisor_cost = 0.0_f64;
    let mut executor_tokens = 0_u64;
    let mut advisor_tokens = 0_u64;
    let mut total_escalations = 0_u64;
    let mut total_calls = 0_u64;
    let mut resolved_count = 0_u64;
    let mut correction_sum = 0_u64;
    let mut models_set = std::collections::HashSet::new();
    let mut model_stats: std::collections::HashMap<String, (u64, u64, f64, String)> =
        std::collections::HashMap::new();

    for sess in &sessions {
        let d = &sess.data;
        let ex_cost = nested_f64(d, "executor_stats", "total_cost_usd");
        let adv_cost = nested_f64(d, "advisor_stats", "total_cost_usd");
        let ex_tokens = nested_u64(d, "executor_stats", "total_tokens");
        let adv_tokens = nested_u64(d, "advisor_stats", "total_tokens");
        let ex_calls = nested_u64(d, "executor_stats", "calls");
        let adv_calls = nested_u64(d, "advisor_stats", "calls");
        let corrections = nested_u64(d, "combined", "user_corrections");

        executor_cost += ex_cost;
        advisor_cost += adv_cost;
        executor_tokens += ex_tokens;
        advisor_tokens += adv_tokens;
        total_escalations += adv_calls;
        total_calls += ex_calls + adv_calls;
        correction_sum += corrections;

        let ex_model = extract_str(d, "executor_model");
        let adv_model = extract_str(d, "advisor_model");
        if !ex_model.is_empty() {
            models_set.insert(ex_model.clone());
            let entry = model_stats.entry(ex_model).or_insert((0, 0, 0.0, "executor".into()));
            entry.0 += ex_calls;
            entry.1 += ex_tokens;
            entry.2 += ex_cost;
        }
        if !adv_model.is_empty() {
            models_set.insert(adv_model.clone());
            let entry = model_stats.entry(adv_model).or_insert((0, 0, 0.0, "advisor".into()));
            entry.0 += adv_calls;
            entry.1 += adv_tokens;
            entry.2 += adv_cost;
        }
    }

    // Also aggregate from pipeline_run packets (real Gemini costs)
    let mut pipeline_cost = 0.0_f64;
    let mut pipeline_tokens = 0_u64;
    for run in &pipeline_runs {
        let d = &run.data;
        if let Some(rc) = d.get("realCost") {
            pipeline_cost += extract_f64(rc, "totalCostUsd");
        }
        if let Some(tu) = d.get("tokenUsage") {
            let total = extract_u64(tu, "totalTokens");
            pipeline_tokens += total;
            let m = extract_str(tu, "model");
            if !m.is_empty() {
                models_set.insert(m.clone());
                let entry = model_stats.entry(m).or_insert((0, 0, 0.0, "pipeline".into()));
                entry.0 += 1;
                entry.1 += total;
                entry.2 += extract_f64(rc_for(d), "totalCostUsd");
            }
        }
    }

    // Count resolved decisions
    for dec in &decisions {
        let d = &dec.data;
        if nested_str(d, "outcome", "").is_empty() {
            if extract_bool(d.get("outcome").unwrap_or(&serde_json::Value::Null), "task_completed") {
                resolved_count += 1;
            }
        }
    }

    let combined = executor_cost + advisor_cost + pipeline_cost;
    let all_tokens = executor_tokens + advisor_tokens + pipeline_tokens;

    // Estimate opus-only cost: all tokens at Opus pricing
    let estimated_opus = (all_tokens as f64 / 1_000_000.0) * 45.0; // blended ~$45/M

    let model_breakdown: Vec<ModelBreakdown> = model_stats
        .into_iter()
        .map(|(model, (calls, tokens, cost, role))| ModelBreakdown {
            model,
            calls,
            total_tokens: tokens,
            total_cost_usd: (cost * 1_000_000.0).round() / 1_000_000.0,
            avg_tokens_per_call: if calls > 0 { tokens / calls } else { 0 },
            role,
        })
        .collect();

    let stats = AdvisorStats {
        total_sessions: sessions.len(),
        total_decisions: decisions.len(),
        total_pipeline_runs: pipeline_runs.len(),
        executor_total_cost_usd: (executor_cost * 1_000_000.0).round() / 1_000_000.0,
        advisor_total_cost_usd: (advisor_cost * 1_000_000.0).round() / 1_000_000.0,
        combined_total_cost_usd: (combined * 1_000_000.0).round() / 1_000_000.0,
        advisor_cost_share_pct: if combined > 0.0 {
            ((advisor_cost / combined) * 1000.0).round() / 10.0
        } else {
            0.0
        },
        executor_total_tokens: executor_tokens,
        advisor_total_tokens: advisor_tokens,
        escalation_rate_pct: if total_calls > 0 {
            ((total_escalations as f64 / total_calls as f64) * 1000.0).round() / 10.0
        } else {
            0.0
        },
        advisor_resolved_pct: if decisions.len() > 0 {
            ((resolved_count as f64 / decisions.len() as f64) * 1000.0).round() / 10.0
        } else {
            0.0
        },
        avg_user_corrections: if sessions.len() > 0 {
            ((correction_sum as f64 / sessions.len() as f64) * 10.0).round() / 10.0
        } else {
            0.0
        },
        estimated_opus_only_cost_usd: (estimated_opus * 100.0).round() / 100.0,
        savings_vs_opus_only_pct: if estimated_opus > 0.0 {
            (((estimated_opus - combined) / estimated_opus) * 1000.0).round() / 10.0
        } else {
            0.0
        },
        models_seen: models_set.into_iter().collect(),
        model_breakdown,
    };

    let msg = if sessions.is_empty() && decisions.is_empty() && pipeline_runs.is_empty() {
        "No advisor data yet. Push advisor.session or advisor.decision packets to see stats.".into()
    } else {
        format!(
            "{} sessions, {} decisions, {} pipeline runs. All costs measured from real API token counts.",
            sessions.len(),
            decisions.len(),
            pipeline_runs.len(),
        )
    };

    Json(StatsResponse { stats, message: msg })
}

/// Helper to safely get realCost from packet data
fn rc_for(data: &serde_json::Value) -> &serde_json::Value {
    data.get("realCost").unwrap_or(&serde_json::Value::Null)
}

async fn list_sessions(
    State(state): State<Arc<AppState>>,
) -> Json<SessionsResponse> {
    state.increment_requests();

    let packets = state.retention_packets.lock().await;
    let sessions: Vec<AdvisorSession> = packets
        .iter()
        .filter(|p| p.packet_type == "advisor.session")
        .map(|p| {
            let d = &p.data;
            AdvisorSession {
                session_id: extract_str(d, "session_id"),
                subject: p.subject.clone(),
                executor_model: extract_str(d, "executor_model"),
                advisor_model: extract_str(d, "advisor_model"),
                executor_tokens: nested_u64(d, "executor_stats", "total_tokens"),
                executor_cost_usd: nested_f64(d, "executor_stats", "total_cost_usd"),
                advisor_tokens: nested_u64(d, "advisor_stats", "total_tokens"),
                advisor_cost_usd: nested_f64(d, "advisor_stats", "total_cost_usd"),
                escalation_count: nested_u64(d, "advisor_stats", "calls"),
                task_completed: nested_str(d, "combined", "task_completed")
                    .parse::<bool>()
                    .unwrap_or(extract_bool(
                        d.get("combined").unwrap_or(&serde_json::Value::Null),
                        "task_completed",
                    )),
                user_corrections: nested_u64(d, "combined", "user_corrections"),
                timestamp: p.timestamp.clone(),
            }
        })
        .collect();

    Json(SessionsResponse { sessions })
}

async fn list_decisions(
    State(state): State<Arc<AppState>>,
) -> Json<DecisionsResponse> {
    state.increment_requests();

    let packets = state.retention_packets.lock().await;
    let decisions: Vec<AdvisorDecision> = packets
        .iter()
        .filter(|p| p.packet_type == "advisor.decision")
        .map(|p| {
            let d = &p.data;
            AdvisorDecision {
                decision_id: extract_str(d, "decision_id"),
                session_id: extract_str(d, "session_id"),
                trigger: extract_str(d, "trigger"),
                advisor_model: nested_str(d, "advisor", "model"),
                advisor_tokens: nested_u64(d, "advisor", "total_tokens"),
                advisor_cost_usd: nested_f64(d, "advisor", "cost_usd"),
                advice_type: nested_str(d, "advisor", "advice_type"),
                advice_summary: nested_str(d, "advisor", "advice_summary"),
                was_applied: extract_bool(
                    d.get("outcome").unwrap_or(&serde_json::Value::Null),
                    "executor_applied",
                ),
                timestamp: extract_str(d, "timestamp"),
            }
        })
        .collect();

    Json(DecisionsResponse { decisions })
}

// ── Route registration ───────────────────────────────────────────────────────

pub fn routes() -> Router<Arc<AppState>> {
    Router::new()
        .route("/stats", get(stats))
        .route("/sessions", get(list_sessions))
        .route("/decisions", get(list_decisions))
}
